File: src/Window.cpp
#include "Spritter/Window.h"

#include <iostream>
#include <stdexcept>

namespace Spritter
{
    Window::Window(const std::string& title, const Math::Size& size)
    {
        if (!SDL_Init(SDL_INIT_VIDEO))
            throw std::runtime_error("Failed to initialize SDL: " + std::string(SDL_GetError()));

        _window = SDL_CreateWindow(title.c_str(), size.Width, size.Height, SDL_WINDOW_OPENGL);
        if (!_window)
            throw std::runtime_error("Failed to create window: " + std::string(SDL_GetError()));
    }

    Window::~Window()
    {
        SDL_DestroyWindow(_window);
        SDL_Quit();
    }
}

File: src/Spritter/Window.cs
using System.Drawing;
using SDL3;

namespace Spritter;

public sealed class Window : IDisposable
{
    private readonly IntPtr _window;

    public IntPtr Handle => _window;

    public Window(string title, Size size)
    {
        if (!SDL.Init(SDL.InitFlags.Video | SDL.InitFlags.Events))
            throw new Exception($"Failed to initialize SDL: {SDL.GetError()}");

        SDL.GLSetAttribute(SDL.GLAttr.ContextMajorVersion, 3);
        SDL.GLSetAttribute(SDL.GLAttr.ContextMinorVersion, 3);
        SDL.GLSetAttribute(SDL.GLAttr.ContextProfileMask, (int) SDL.GLProfile.Core);
        
        _window = SDL.CreateWindow(title, size.Width, size.Height, SDL.WindowFlags.OpenGL);

        if (_window == IntPtr.Zero)
            throw new Exception($"Failed to create window: {SDL.GetError()}");
    }
    
    public void Dispose()
    {
        SDL.DestroyWindow(_window);
        SDL.Quit();
    }
}